### PR TITLE
Fix execution of CompilationStartActions when only analyzer options for the project change, but the underlying compilation is identical.

### LINKED ISSUE #530: Changes to additional files sometimes don't lead to diagnostic re-computation in IDE
Repro steps:
1) Open Roslyn.sln in RoslynDev hive after building Roslyn
2) Open PublicAPI.txt file from Workspaces.Desktop project
3) Open FileTextLoader.cs and wait for few minutes (so that its diagnostics are computed)
4) Delete the 4 entries for FileTextLoader at the top of PublicAPI.txt
5) Verify that squiggles show up in FileTextLoader for RS0016
6) Go back to PublicAPI.txt and undo the changes in step (4)

Expected: Squiggles in FileTextLoader.cs go away.
Actual: Squiggles in FileTextLoader.cs never go away

### PR DESCRIPTION
**Pull request by mavasani**
Our current assumption for executing CompilationStartActions is that they should be executed exactly once per-compilation per-analyzer. We cache the registered nested actions for every compilation in AnalyzerManager to ensure this. However, if user just changes an additional document for the project, then the analyzer options change, but the underlying compilation is unchanged. Project change causes IDE diagnostic service to clear all cached diagnostics, but when re-computing diagnostics, we end up skipping execution of CompilationStartActions and directly move to nested compilation actions. This can be troublesome in couple of ways. First, we will have duplicate callbacks for all nested actions for the initially invoked compilation start action. Second, and event worse, if the analyzer is reading the contents of the additional documents in the start action and caching them for later use by the nested actions, all the nested actions get stale contents for the additional documents. Apparently, Tom is the first one to go through this agony as the DeclarePublicAPI analyzer that he implemented exactly does the latter.

Fix is to ensure that we execute compilation start actions for every unique compilation start analysis context, i.e. {compilation, analyzer options} tuple for every analyzer. This change fixes #530 

@tmeschter @srivatsn @heejaechang @JohnHamby can you please review?

@JohnHamby this is something that we may want to make it explicit in the analyzer API semantics documentation?

### CODE REVIEW

File: src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs [hunk 1]
```diff
@@ -37,9 +35,13 @@ internal class AnalyzerManager
...
-            new ConditionalWeakTable<Compilation, ConcurrentDictionary<DiagnosticAnalyzer, Task<HostCompilationStartAnalysisScope>>>();
+        // Compilation start actions will get executed once per-each AnalyzerAndOptions as user might want to return different set of custom actions for each compilation/analyzer options.
+        private readonly ConditionalWeakTable<Compilation, ConcurrentDictionary<AnalyzerAndOptions, Task<HostCompilationStartAnalysisScope>>> _compilationScopeMap =
+            new ConditionalWeakTable<Compilation, ConcurrentDictionary<AnalyzerAndOptions, Task<HostCompilationStartAnalysisScope>>>();
+
+        private readonly ConditionalWeakTable<Compilation, ConcurrentDictionary<AnalyzerAndOptions, Task<HostCompilationStartAnalysisScope>>>.CreateValueCallback _compilationScopeMapCallback =
+            new ConditionalWeakTable<Compilation, ConcurrentDictionary<AnalyzerAndOptions, Task<HostCompilationStartAnalysisScope>>>.CreateValueCallback(
+                comp => new ConcurrentDictionary<AnalyzerAndOptions, Task<HostCompilationStartAnalysisScope>>(AnalyzerAndOptions.Comparer));
```
Comment (heejaechang) [inline src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs]: do you need comparer here? you implemented Equals and HashCode in AnalyzerAndOptions.

and we probably want to limit concurrency level on ConcurrentDictionary.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerOptions.cs [hunk 2]
```diff
@@ -37,5 +38,17 @@ public AnalyzerOptions WithAdditionalFiles(ImmutableArray<AdditionalText> additi
 
             return new AnalyzerOptions(additionalFiles);
         }
+
+        public override bool Equals(object obj)
+        {
+            var other = obj as AnalyzerOptions;
```
Comment (heejaechang) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerOptions.cs]: can we check whether two AnalyzerOptions has same reference equality first and then see whether AdditionalFiles has same and then each items.

**Comment by heejaechang** [comment]
:+1:

**Comment by tmeschter** [comment]
LGTM, other than the test failure in Jenkins.

**Comment by mavasani** [comment]
@heejaechang  @tmeschter I have addressed the test failure and review feedback in the new iteration. Test.cmd passes locally for me after this iteration.

**Comment by srivatsn** [comment]
:+1:

**Pull request merged**